### PR TITLE
feat(GitHub): add file create/update to Repo Operations

### LINKED ISSUE #2023: Add `Contents > Create or update file contents` operation to the OOTB GitHub Connector
__Is your feature request related to a problem? Please describe.__

If I want to create or update GitHub file contents from my C8 process, I have to manually put in all necessary data using the REST Connector to call https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents. This is quite error-prone since I have to know all necessary headers and define the request body correctly. 

Additionally, the GitHub API requires the file content passed in the request body to be `Base64` encoded which requires cumbersome workarounds if the content is not already encoded when receiving it from another source (e.g. Web Modeler's REST API, see https://modeler.cloud.camunda.io/swagger-ui/index.html#/Files/getFile).

__Describe the solution you'd like__

Using the GitHub connector, I would like to have the option to create or update file contents.
I would want to input as little data as possible and as much custom data as required. 
The file content passed in the request payload should be automatically `Base64` encoded.

__Describe alternatives you've considered__

Use the REST Connector and define my own template once I figured out all the necessary headers and how to define the body correctly.

__Additional context__

- API to run create or update file contents: https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents
- Result of https://github.com/camunda/web-modeler/issues/7842

### PR DESCRIPTION
**Pull request by vringar**
## Description

This feature adds the ability to create or update files in a repository using [this endpoint](https://docs.github.com/en/rest/repos/contents?apiVersion=2022-11-28#create-or-update-file-contents).

Features that I've covered:
- Full coverage of API including optional author 
- User provided content can either be base64 encoded or not

## Related issues

Closes #2023

## Checklist

- [x] PR has a **milestone** or the `no milestone` label.
- [x] Docs PR [is open](https://github.com/camunda/camunda-docs/pull/5979)

### CODE REVIEW

File: connectors/github/element-templates/github-connector.json [hunk 1]
```diff
@@ -742,6 +747,22 @@
         ]
       }
     },
+    {
+      "id": "reposMethodPut",
```
Comment (sbuettner) [inline connectors/github/element-templates/github-connector.json:751]: Maybe `uploadFile`?

File: connectors/github/element-templates/github-connector.json [hunk 2]
```diff
@@ -2279,6 +2561,22 @@
       },
       "optional": true
     },
+    {
+      "group": "input",
+      "type": "Hidden",
+      "value": "=context merge({\"message\": commitMessage, \"content\": b64content}, if sha != null then {\"sha\": sha} else {}, if branch != null then {\"branch\": branch} else {}, if committerName != null and committerEmail != null then {\"committer\": {\"name\": committerName, \"email\": committerEmail, \"date\": committerDate}} else {}, if authorName != null and authorEmail != null then {\"author\": {\"name\": authorName, \"email\": authorEmail, \"date\": authorDate }} else {})",
```
Comment (vringar) [inline connectors/github/element-templates/github-connector.json:2567]: If the branch is set to null instead of being absent, github will return a 422, so I had to get a bit creative here.
Idk if we prefer this pattern over `context from (get entries(context)[value != null])
I couldn't find either pattern in the codebase

**Review by sbuettner — APPROVED** [review]
Congrats to your first PR! 🎉 

Looks great to me. I would invite @johnBgood for a second pair of eyes.

If you dont find any issues testing feel free to merge it :)

In this case (as we dont break compatibility we dont need to increase the version. In general I recommend to take a look at the element template lifecycle: https://github.com/bpmn-io/element-templates/blob/main/docs/LIFE_CYCLE.md#life-cycle

File: connectors/github/element-templates/github-connector.json [hunk 3]
```diff
@@ -2184,6 +2208,264 @@
...
+      },
+      "condition": {
+        "property": "useCustomCommitter",
+        "equals": true,
+        "type": "simple"
+      }
+    },
+    {
+      "label": "Commit date",
+      "description": "The date of the commit, using ISO-8601",
```
Comment (johnBgood) [inline connectors/github/element-templates/github-connector.json:2382]: Maybe can we use the `tooltip` property to give an example date as it's always annoying to find out (or maybe a documentation link)

File: connectors/github/element-templates/github-connector.json [hunk 4]
```diff
@@ -2184,6 +2208,264 @@
...
+      "condition": {
+        "property": "useCustomAuthor",
+        "equals": true,
+        "type": "simple"
+      }
+    },
+
+    {
+      "label": "Author date",
+      "description": "The authoring date, using ISO-8601",
```
Comment (johnBgood) [inline connectors/github/element-templates/github-connector.json:2455]: same here

**Review by johnBgood — COMMENTED** [review]
Nice one! Just left some questions/comments :) 

EDIT: @vringar when this is ready to be tested, you can add the `qa:required` label to the PR. It will post a comment that you can edit, adding information for @Szik to test the PR. 
We can discuss the process if you want to.

File: connectors/github/element-templates/github-connector.json [hunk 5]
```diff
@@ -2184,6 +2208,264 @@
...
+      "condition": {
+        "property": "reposOperationType",
+        "oneOf": [
+          "uploadFile"
+        ]
+      }
+    },
+    {
+      "label": "Content",
+      "description": "The file to be uploaded. Please check the checkbox below, if it is already base64 encoded",
```
[the remark this answers is not in the document]
Comment (vringar) [reply connectors/github/element-templates/github-connector.json:2281]: @johnBgood so we drop the case of already b64 encoded content? That would would allow us to drop the checkbox below, but since we don't provide a way to decode content (e.g. go from b64 string to document) I'm unsure if we can just drop that part.

File: connectors/github/element-templates/github-connector.json (hunk 3 above)
Comment (vringar) [reply connectors/github/element-templates/github-connector.json:2382]: My current problem is that I haven't figured out what the correct format is yet. According to the docs 2023-10-01T12:00:00Z should work but it doesn't. I'll have to experiment some more.

**Comment by vringar** [comment]
Waiting on Design input for the following decision:

How do we best model the optional committer and author inputs. Both of these inputs have the same structure:
Name and email need to be provided together, with an additional optional date.

Desired Outcome:
1. Either both the name or the email are supplied but not just one or the other
2. The date can only be specified if both other fields are present
3. All of the fields can be set dynamically via FEEL

Option 1 - Checkbox with required fields:

1. Achieved by making both fields required when the checkbox is set
2. Date field is only visible when checkbox is set
3. Doesn't work, because we can't show the fields based on the absence of the empty checkbox (to include both true and free text FEEL). See https://github.com/camunda/camunda-modeler/issues/5087 for more context

Option 2 - Combine name and email field into one and make it always visible and optional

1. Can use validation to ensure both pieces of information are present in the string like "Name <email@example.com>"
2. Date field can be set without setting author. Would be silently ignored since we can't warn on this
3. Full FEEL support as we don't have any hidden fields

Option 3 - Keep fields separate and always show them (For completeness, seems worse than Option 2)

1. Not possible
2. Not possible
3. Full FEEL support

**Comment by johnBgood** [comment]
# QA information
## Test Environment
- Test required in (you can check both):
  - [ ] SaaS
  - [ ] SM
- [ ] Create a new test for this feature in our [Pre Release Tests](https://modeler.ultrawombat.com/projects/017779a8-9459-4feb-8b47-eeef8b2717ef--connectors-pre-release-test) for @Szik to be able to test

## Test Scope
*Please describe the test scope, happy path, edge cases that come to your mind, and whatever you think might relevant to test*

## Test Data
*Please provide the test data, if needed (files, URLs, code snippets, FEEL expressions)*

**Comment by Szik** [comment]
open questions form QA:
- "Create of update File" > "SHA" mandatory? (file tagged as not required; assumed default to be picked; atm: manually looked up > too cumbersome ) 
- label "branch": "The name of the branch. Default: the repository's main branch" > "The name of the branch. Default value: main" (or somthing like that
- label "Path": "The path of the content within the repository" > "The path of the file within the repository"

**Comment by vringar** [comment]
> "Create of update File" > "SHA" mandatory? (file tagged as not required; assumed default to be picked; atm: manually looked up > too cumbersome )

The SHA is only mandatory if a file already exists at the given location. As we are only wrapping the REST connector, we can't make multiple calls "under the hood" to see if the file exists and provide the SHA as appropriate.

> label "branch": "The name of the branch. Default: the repository's main branch" > "The name of the branch. Default value: main" (or somthing like that

The default branch of a repository is [configurable](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch). 
I copied the description from [here](https://github.com/camunda/connectors/blob/2d072c519ca972b54e4d5d764e420a7693471968/connectors/github/element-templates/github-connector.json#L2170) for consistencies sake.

Should I change both to default branch?

> label "Path": "The path of the content within the repository" > "The path of the file within the repository"

Fixed :) Note that this also changes it for "Get repository content"

**Review by sbuettner — APPROVED** [review]
